force_bdss/app/tests/test_evaluate_operation.py

```diff
@@ -100,13 +105,33 @@ def test_error_for_non_matching_mco_parameters(self):
         self.operation.workflow.mco_model = factory.create_model()
 
         with testfixtures.LogCapture():
-            with self.assertRaisesRegex(
+            with (self.assertRaisesRegex(
+                    RuntimeError,
+                    "Workflow file has errors.") or
+                  self.assertRaisesRegex(
                     RuntimeError,
                     r"The number of data values returned by the MCO "
                     r"\(1 values\) does not match the number of "
                     r"parameters specified \(0 values\). This is either "
                     "a MCO plugin error or the workflow file is "
-                    "corrupted."):
+                    "corrupted.")):
+                self.operation.run()
+
+    def test_error_for_missing_ds_output_names(self):
+
+        factory = self.registry.data_source_factories[0]
+        factory.output_slots_size = 2
+        self.operation.workflow_file.read()
+
+        with testfixtures.LogCapture():
+            with (self.assertRaisesRegex(
+                    RuntimeError,
+                    "Workflow file has errors.") or
+                  self.assertRaisesRegex(
+                    RuntimeError,
+                    r"The number of data values \(2 values\)"
+                    " returned by 'test_data_source' does not match"
+                    " the number of user-defined names")):
```

Comment:
I wouldn't expect this error statement to be returned - are you able to isolate when it would be raised?

Reply:
This is in the master branch and is at least part of a error message raised by ExecutionLayer.execute_layer (line 101, execution_layer.py).
I took it out and coverage works (locally) as well as codecov/patch. But codecov/project fails Whether it is there or not.

Reply:
In the latest commit (bc271c0..) I just removed the entire test. As the error message didn't completely match maybe it was vestigial (tests an error no longer there?). There is no equivalent test for the optimize operation.

Reply:
I think this happens because in `verify_workflow` we log the individual error messages before raising a new generic error (the `Workflow file has errors` message). Therefore all of the individual errors are no longer going to be reported through `assertRaisesRegex`. In which case we can check the log for these individual messages, rather than the traceback.

The following seems to work for `test_error_for_missing_ds_output_names` (I haven't checked if it improves the coverage)

```python
        with testfixtures.LogCapture() as capture:
            with (self.assertRaisesRegex(
                    RuntimeError,
                    "Workflow file has errors.")):
                self.operation.run()
            capture.check(
                ('force_bdss.app.base_operation',
                 'ERROR',
                 'Unable to execute workflow due to verification errors:'),
                ('force_bdss.app.base_operation',
                 'ERROR',
                 'The number of output slots is incorrect.')
            )
```

Reply:
> This is in the master branch and is at least part of a error message raised by ExecutionLayer.execute_layer (line 101, execution_layer.py).
> I took it out and coverage works (locally) as well as codecov/patch. But codecov/project fails Whether it is there or not.

Ohhhh, I see what's going on now. Ok - this is worth a separate PR in it's own right. It looks like there's two places where the input / output slot lengths are being checked - in the `BaseDataSourceModel.verify` method (which is called by `verify_workflow`) and also as you spotted in `ExecutionLayer.execute_layer`.

These means we would get 2 different error messages for the same error, depending on which method raised it - not ideal. We should make sure that only goes on in one place.